segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/ZKSegmentContainerMonitor.java

```diff
@@ -106,9 +107,16 @@ public void initialize() {
     public void initialize(Duration monitorInterval) {
         Exceptions.checkNotClosed(closed.get(), this);
 
+        // Start loading the segment container to node assigment map from zookeeper.
         this.hostContainerMapNode.start();
+
+        // There are two triggers for the segment container monitor.
+        // 1. On any segment container ownership changes notified via zookeeper. We will ensure the local containers
+        //      are stopped/started according to the new ownership mapping.
+        // 2. At scheduled intervals to perform retries on local segment container start failures.
         this.assigmentTask.set(this.executor.scheduleWithFixedDelay(
                 this::checkAssignment, 0L, monitorInterval.getSeconds(), TimeUnit.SECONDS));
+        this.hostContainerMapNode.getListenable().addListener(this::checkAssignment, this.executor);
```

Comment:
I had initially started with a queuing approach (similar to the controller equivalent for this in SegmentMonitorLeader -
 in the controller's case we already had a dedicated thread hence implemented this using semaphores for queuing the notifications) but couldn't implement it without having a dedicated thread blocked for this purpose. In the current implementation the checkAssignment is triggered only when necessary and does not need a dedicated thread. Also checkAssignment only does an inmemory check and does not use any blocking APIs, so even though there could be potential lock contention it should not have any significant blocking effect. Also since the controller in practice has heuristics to ensure that the assignment is not modified in quick successions, the only time contention can happen is when the regular schedule and notification from ZK is triggered at the same time, and this was the reason we put this function under a lock.